cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -156,7 +163,11 @@ struct IterationTraits<fs::FileInfoVector> {
 namespace fs {
 
 /// \brief Abstract file system API
-class ARROW_EXPORT FileSystem : public std::enable_shared_from_this<FileSystem> {
+class ARROW_EXPORT FileSystem
+    /// \cond false
+    : public std::enable_shared_from_this<FileSystem>
+/// \endcond
+{  // NOLINT
  public:
   virtual ~FileSystem();
 
```

Comment:
After all filesystems are hidden behind FileSystemFromUri, it might be desirable to supplement or replace `type_name()` with a member which returns a URI scheme that corresponds to the factory used to create the instance. Taking this a step further, we could also add full round tripping of filesystems through URI strings. Both of these seem potentially useful for interactive use cases where a user might receive a filesystem and want to know how it was constructed.